src/protocol/codecs/algomsg.rs

```diff
@@ -0,0 +1,155 @@
+use std::io;
+
+use bytes::BytesMut;
+use tokio_util::codec::Decoder;
+use tracing::*;
+
+use super::topic::TopicCodec;
+use crate::protocol::invalid_data;
+
+/// [Tag] represents a message type identifier.
+///
+/// The original tag list can be found in go-algorand/protocol/tags.go.
+#[derive(Debug)]
+pub enum Tag {
+    UnknownMsg,
+    AgreementVote,
+    MsgOfInterest,
+    MsgDigestSkip,
+    NetPrioResponse,
+    Ping,
+    PingReply,
+    ProposalPayload,
+    StateProofSig,
+    TopicMsgResp,
+    Txn,
+    UniCatchupReq,
+    UniEnsBlockReq,
+    VoteBundle,
+}
+
+impl TryFrom<BytesMut> for Tag {
+    type Error = io::Error;
+
+    fn try_from(tag: BytesMut) -> Result<Self, Self::Error> {
```

Comment:
```suggestion
    fn try_from(tag: Bytes) -> Result<Self, Self::Error> {
```
unless it's an API annoyance

Reply:
yep, it doesn't work...

Reply:
what if you apply `.freeze()` at the callsite? I mean
```
let tag = Tag::try_from(src.split_to(TAG_LEN).freeze())?;
```